sql/expression/function/format.go

```diff
@@ -0,0 +1,183 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package function
+
+import (
+	"fmt"
+	"math"
+	"strings"
+
+	"golang.org/x/text/language"
+	"golang.org/x/text/message"
+
+	"github.com/dolthub/go-mysql-server/sql"
+)
+
+// Format function returns a result of numX rounded to df decimal places as a string.
+type Format struct {
+	Val    sql.Expression
+	Df     sql.Expression
+	Locale sql.Expression
+}
+
+var _ sql.FunctionExpression = (*Format)(nil)
+
+// NewFormat returns a new Format expression.
+func NewFormat(args ...sql.Expression) (sql.Expression, error) {
+	var numX, df, locale sql.Expression
+	switch len(args) {
+	case 2:
+		numX = args[0]
+		df = args[1]
+		locale = nil
+	case 3:
+		numX = args[0]
+		df = args[1]
+		locale = args[2]
+	default:
+		return nil, sql.ErrInvalidArgumentNumber.New("FORMAT", "2 or 3", len(args))
+	}
+	return &Format{numX, df, locale}, nil
+}
+
+// FunctionName implements sql.FunctionExpression
+func (f *Format) FunctionName() string {
+	return "format"
+}
+
+// Type implements the Expression interface.
+func (f *Format) Type() sql.Type { return sql.LongText }
+
+// IsNullable implements the Expression interface.
+func (f *Format) IsNullable() bool {
+	return f.Val.IsNullable() || f.Df.IsNullable() || f.Locale.IsNullable()
+}
+
+func (f *Format) String() string {
+	return fmt.Sprintf("format(%s, %s, %s)", f.Val, f.Df, f.Locale)
+}
+
+// Eval implements the Expression interface.
+func (f *Format) Eval(ctx *sql.Context, row sql.Row) (interface{}, error) {
+	value, err := f.Val.Eval(ctx, row)
+	if err != nil {
+		return nil, err
+	}
+	if value == nil {
+		return nil, nil
+	}
+
+	if !sql.IsNumber(f.Val.Type()) {
+		value, err = sql.Float64.Convert(value)
+		if err != nil {
+			return nil, nil
+		}
+	}
+
+	var dVal float64
+	dTemp, err := f.Df.Eval(ctx, row)
+	if err != nil {
+		return nil, err
+	}
+	if dTemp == nil {
+		return nil, nil
+	}
+
+	switch dNum := dTemp.(type) {
+	case float64:
+		dVal = float64(int64(math.Round(dNum)))
+	case float32:
+		dVal = float64(int64(math.Round(float64(dNum))))
+	case int64:
+		dVal = float64(dNum)
+	case int32:
+		dVal = float64(dNum)
+	case int16:
+		dVal = float64(dNum)
+	case int8:
+		dVal = float64(dNum)
+	case uint64:
+		dVal = float64(dNum)
+	case uint32:
+		dVal = float64(dNum)
+	case uint16:
+		dVal = float64(dNum)
+	case uint8:
+		dVal = float64(dNum)
+	case int:
+		dVal = float64(dNum)
+	default:
+		dTemp, err = sql.Float64.Convert(dTemp)
+		if err == nil {
+			dVal = dTemp.(float64)
+		} else {
+			return nil, nil
+		}
+	}
+
+	if dVal < float64(0) {
+		dVal = float64(0)
+	}
+	xVal := math.Round(value.(float64)*math.Pow(10.0, dVal)) / math.Pow(10.0, dVal)
```

Comment:
This and the below is a weird way to split a float in to whole and fractional pieces

Do this instead:
https://www.techieindoor.com/go-how-to-get-integer-and-fractional-part-of-floating-point-numbers-in-go/